src/workerd/api/actor.h

```diff
@@ -13,27 +13,101 @@
 #include <capnp/compat/http-over-capnp.h>
 #include <workerd/api/http.h>
 #include <workerd/jsg/jsg.h>
+#include <workerd/jsg/function.h>
 
 namespace workerd {
   template <typename T> class IoOwn;
 }
 
 namespace workerd::api {
 
+enum IsInHouse {
+  YES,
+  NO,
+};
+
+// This class is the client RPC stub for Durable Objects. It intercepts dynamic method calls, then
+// sends the method name and arguments to the remote Durable Object, returning the response it
+// receives.
+//
+// We don't want name interception on Fetcher itself -- at least not yet -- since not all Fetcher
+// implementations need it. Name interception is off by default; to turn it on, set the JsRpcStub
+// compatibility flag.
+class DurableObjectRpcStub : public Fetcher, public jsg::NamedIntercept {
+public:
+  DurableObjectRpcStub(
+      IoOwn<OutgoingFactory> outgoingFactory,
+      RequiresHostAndProtocol requiresHost,
+      IsInHouse inHouse,
+      CompatibilityFlags::Reader flags)
+    : Fetcher(kj::mv(outgoingFactory), requiresHost, (inHouse == YES)), flags(flags) {}
+
+  bool isRegisteredName(kj::StringPtr name) {
+    // We don't need to check `fetch` or `constructor` because `fetch` is known at compile time and
+    // `constructor` is a JS keyword; neither uses name interception. This needs to be updated
+    // whenever a new "registered name", aka handler, is added to Durable Objects.
+    if (name == "alarm"_kjc ||
+        name == "webSocketMessage"_kjc ||
+        name == "webSocketClose"_kjc ||
+        name == "webSocketError"_kjc) {
+      return true;
+    }
+    return false;
+  }
+
+  kj::Maybe<jsg::JsValue> getNamed(jsg::Lock& js, kj::StringPtr name) override {
+    if (flags.getJsRpcStub()) {
+      JSG_REQUIRE(!isRegisteredName(name), Error, "cannot call registered method '", name, "'");
+
+      // Named intercept is enabled, this means we won't default to legacy behavior.
+      // TODO(soon): The server side PR will fill this function with the RPC call and other stuff,
+      //             this is just a temporary placeholder for testing.
+      return jsg::JsValue(js.wrapReturningFunction(js.v8Context(),
+          [](jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
+            return js.str("foo"_kj);
```

Comment:
Hmmmmm, I'm thinking maybe we should just return something like `js.error("JsRpcStub is unimplemented")`. Then we can land this without rebasing on the server side branch for now.